Change `settings` to `payload` and use google protobuf Struct

stefan-brus-frequenz opened the pull request.

stefan-brus-frequenz committed:
Clear release notes

Signed-off-by: Stefan Brus <stefan.brus@frequenz.com>

stefan-brus-frequenz committed:
Revert dispatch settings to use protobuf `Struct`

The dispatch API should not care about any settings that are used by
downstream apps.

Signed-off-by: Stefan Brus <stefan.brus@frequenz.com>

stefan-brus-frequenz committed:
Rename settings field to `payload`

Make it clearer that this field contains some arbitrary data from the
API's POV.

Signed-off-by: Stefan Brus <stefan.brus@frequenz.com>

stefan-brus-frequenz committed:
Update release notes

Signed-off-by: Stefan Brus <stefan.brus@frequenz.com>

stefan-brus-frequenz committed:
Remove fcr folder from python

Signed-off-by: Stefan Brus <stefan.brus@frequenz.com>

stefan-brus-frequenz committed:
Merge branch 'v0.x.x' into settings_payload

Signed-off-by: stefan-brus-frequenz <105632599+stefan-brus-frequenz@users.noreply.github.com>

stefan-brus-frequenz committed:
Remove "fcr pq import" test

Signed-off-by: Stefan Brus <stefan.brus@frequenz.com>

llucax reviewed (dismissed):
Not sure if we are 100% decided about this already, but LGTM.

tiyash-basu-frequenz commented:
Do we know if protobuf structs are what we want? What if we use json configs as a standard?

stefan-brus-frequenz commented:
Putting a blocked on this until after the meeting

stefan-brus-frequenz commented:
> Do we know if protobuf structs are what we want? What if we use json configs as a standard?

As far as I can tell, protobuf structs are the de facto standard way of representing JSON objects

stefan-brus-frequenz committed:
Remove DispatchType enum and use a string field in its place

Signed-off-by: Stefan Brus <stefan.brus@frequenz.com>

stefan-brus-frequenz committed:
Replace DispatchStatus enum with "active" and "dry run" states

Signed-off-by: Stefan Brus <stefan.brus@frequenz.com>

tiyash-basu-frequenz commented on proto/frequenz/api/dispatch/dispatch.proto:
```diff
@@ -127,7 +99,7 @@ message DispatchFilter {
 
   // Filter by dispatch type
-  repeated DispatchType types = 3;
+  repeated string types = 3;
```
I'd suggest adding documentation about what it is meant for, now that the enum is being removed.

tiyash-basu-frequenz commented on proto/frequenz/api/dispatch/dispatch.proto:
```diff
@@ -131,11 +122,14 @@ message Dispatch {
+  bool active = 8;
+
+  // The "dry run" status
+  bool dry_run = 9;
```
Since these are boolean flags now, then maybe `is_active` and `is_dry_run` are more conventional here? Not a strong opinion though.

tiyash-basu-frequenz commented on proto/frequenz/api/dispatch/dispatch.proto:
```diff
@@ -191,9 +188,12 @@ message DispatchUpdateRequest {
-  // The dispatch status
-  optional DispatchStatus status = 8;
+  // The "active" status
+  optional bool status = 8;
```
I'd suggest being more consistent here, and go with `active` (or `is_active`), instead of `status`.

stefan-brus-frequenz replied:
Oops, missed this one.

tiyash-basu-frequenz commented on proto/frequenz/api/dispatch/dispatch.proto:
```diff
@@ -90,7 +90,7 @@ message DispatchFilter {
 
   // Filter by time interval
-  TimeIntervalFilter time_interval = 5;
+  optional TimeIntervalFilter time_interval = 5;
```
Marking non-primitive data-members as `optional` is redundant. Doing this leads readers to other rabbit holes, like if something is not marked `optional`, should it be marked as `required`? How do I read a non-primitive type that is neither `optional` nor `required`? Since non-primitive types are optional by default in proto3, I think we should not do this :)

stefan-brus-frequenz committed:
Prefix status fields with "is_" for clarity

Signed-off-by: Stefan Brus <stefan.brus@frequenz.com>

stefan-brus-frequenz committed:
Remove "optional" qualifier from non-scalar fields

Signed-off-by: Stefan Brus <stefan.brus@frequenz.com>

stefan-brus-frequenz committed:
Update dispatch type documentation

Signed-off-by: Stefan Brus <stefan.brus@frequenz.com>

stefan-brus-frequenz committed:
Move "Dispatch" message definition to the top

This is the main "object" produced by the API, so it has been moved near to
the top of the definitions file, for visibility.

Signed-off-by: Stefan Brus <stefan.brus@frequenz.com>

thomas-nicolai-frequenz reviewed (commented):
Looks decent to me but I will need to check the entire protobuf again to see if everything is included we need. One thing that might be missing is the last_update_time of a dispatch request. Right now we only reveal the create_time but that will not be enough of caching purposes on the receiving end imho.
